File: implementations/go/examples/internal/infrastructure/persistence/order_repository.go
package persistence

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/example/order-service/internal/domain/order"
)

type OrderRepository struct {
	db *sql.DB
}

// 컴파일 타임 interface 충족 검증
var _ order.Repository = (*OrderRepository)(nil)

func NewOrderRepository(db *sql.DB) *OrderRepository {
	return &OrderRepository{db: db}
}

func (r *OrderRepository) FindByID(ctx context.Context, orderID string) (*order.Order, error) {
	row := r.db.QueryRowContext(ctx,
		`SELECT order_id, user_id, status FROM orders WHERE order_id = $1 AND deleted_at IS NULL`,
		orderID,
	)
	var id, userID, status string
	if err := row.Scan(&id, &userID, &status); err != nil {
		if err == sql.ErrNoRows {
			return nil, order.ErrNotFound
		}
		return nil, fmt.Errorf("find order by id: %w", err)
	}
	items, err := r.findItems(ctx, orderID)
	if err != nil {
		return nil, err
	}
	return order.Reconstitute(id, userID, items, order.Status(status)), nil
}

func (r *OrderRepository) FindAll(ctx context.Context, q order.FindQuery) ([]*order.Order, int, error) {
	args := []any{}
	where := []string{"o.deleted_at IS NULL"}
	i := 1

	if q.UserID != "" {
		where = append(where, fmt.Sprintf("o.user_id = $%d", i))
		args = append(args, q.UserID)
		i++
	}
	if len(q.Status) > 0 {
		placeholders := make([]string, len(q.Status))
		for j, s := range q.Status {
			placeholders[j] = fmt.Sprintf("$%d", i)
			args = append(args, s)
			i++
		}
		where = append(where, fmt.Sprintf("o.status IN (%s)", strings.Join(placeholders, ",")))
	}

	whereClause := strings.Join(where, " AND ")

	var total int
	if err := r.db.QueryRowContext(ctx,
		fmt.Sprintf(`SELECT COUNT(*) FROM orders o WHERE %s`, whereClause), args...,
	).Scan(&total); err != nil {
		return nil, 0, fmt.Errorf("count orders: %w", err)
	}

	args = append(args, q.Take, q.Page*q.Take)
	rows, err := r.db.QueryContext(ctx,
		fmt.Sprintf(`SELECT o.order_id, o.user_id, o.status FROM orders o WHERE %s ORDER BY o.order_id DESC LIMIT $%d OFFSET $%d`, whereClause, i, i+1),
		args...,
	)
	if err != nil {
		return nil, 0, fmt.Errorf("find orders: %w", err)
	}
	defer rows.Close()

	var orders []*order.Order
	for rows.Next() {
		var id, userID, status string
		if err := rows.Scan(&id, &userID, &status); err != nil {
			return nil, 0, err
		}
		items, err := r.findItems(ctx, id)
		if err != nil {
			return nil, 0, err
		}
		orders = append(orders, order.Reconstitute(id, userID, items, order.Status(status)))
	}
	return orders, total, rows.Err()
}

func (r *OrderRepository) Save(ctx context.Context, o *order.Order) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO orders (order_id, user_id, status, updated_at)
		 VALUES ($1, $2, $3, NOW())
		 ON CONFLICT (order_id) DO UPDATE SET status = EXCLUDED.status, updated_at = NOW()`,
		o.OrderID, o.UserID, string(o.Status),
	)
	if err != nil {
		return fmt.Errorf("save order: %w", err)
	}
	return nil
}

func (r *OrderRepository) Delete(ctx context.Context, orderID string) error {
	now := time.Now()
	_, err := r.db.ExecContext(ctx,
		`UPDATE orders SET deleted_at = $1 WHERE order_id = $2 AND deleted_at IS NULL`,
		now, orderID,
	)
	if err != nil {
		return fmt.Errorf("delete order: %w", err)
	}
	return nil
}

func (r *OrderRepository) findItems(ctx context.Context, orderID string) ([]order.Item, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT item_id, name, price, quantity FROM order_items WHERE order_id = $1 AND deleted_at IS NULL`,
		orderID,
	)
	if err != nil {
		return nil, fmt.Errorf("find order items: %w", err)
	}
	defer rows.Close()

	var items []order.Item
	for rows.Next() {
		var item order.Item
		if err := rows.Scan(&item.ItemID, &item.Name, &item.Price, &item.Quantity); err != nil {
			return nil, err
		}
		items = append(items, item)
	}
	return items, rows.Err()
}

File: implementations/fastapi/examples/src/order/infrastructure/persistence/order_repository.py
from datetime import datetime

from sqlalchemy import delete, select, update, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

from ...domain.order import Order, ItemInput, OrderStatus
from ...domain.order_item import OrderItem
from ...domain.repository import OrderRepository


class Base(DeclarativeBase):
    pass


class OrderModel(Base):
    __tablename__ = "orders"

    order_id: Mapped[str] = mapped_column(primary_key=True)
    user_id: Mapped[str]
    status: Mapped[str]
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)
    deleted_at: Mapped[datetime | None] = mapped_column(nullable=True, default=None)


class OrderItemModel(Base):
    __tablename__ = "order_items"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    order_id: Mapped[str]
    item_id: Mapped[int]
    name: Mapped[str]
    price: Mapped[int]
    quantity: Mapped[int]
    deleted_at: Mapped[datetime | None] = mapped_column(nullable=True, default=None)


class SqlAlchemyOrderRepository(OrderRepository):

    def __init__(self, session: AsyncSession) -> None:
        self._session = session

    async def find_by_id(self, order_id: str) -> Order | None:
        stmt = select(OrderModel).where(
            OrderModel.order_id == order_id,
            OrderModel.deleted_at.is_(None),
        )
        row = (await self._session.execute(stmt)).scalar_one_or_none()
        if row is None:
            return None
        items = await self._load_items(order_id)
        return self._to_domain(row, items)

    async def find_all(
        self,
        page: int,
        take: int,
        user_id: str | None = None,
        status: list[str] | None = None,
    ) -> tuple[list[Order], int]:
        stmt = select(OrderModel).where(OrderModel.deleted_at.is_(None))
        count_stmt = select(func.count()).select_from(OrderModel).where(OrderModel.deleted_at.is_(None))

        if user_id:
            stmt = stmt.where(OrderModel.user_id == user_id)
            count_stmt = count_stmt.where(OrderModel.user_id == user_id)
        if status:
            stmt = stmt.where(OrderModel.status.in_(status))
            count_stmt = count_stmt.where(OrderModel.status.in_(status))

        total = (await self._session.execute(count_stmt)).scalar_one()
        rows = (await self._session.execute(
            stmt.order_by(OrderModel.order_id.desc()).offset(page * take).limit(take)
        )).scalars().all()

        orders = []
        for row in rows:
            items = await self._load_items(row.order_id)
            orders.append(self._to_domain(row, items))

        return orders, total

    async def save(self, order: Order) -> None:
        existing = await self._session.get(OrderModel, order.order_id)
        if existing:
            existing.status = order.status
            existing.updated_at = datetime.utcnow()
        else:
            self._session.add(OrderModel(
                order_id=order.order_id,
                user_id=order.user_id,
                status=order.status,
            ))
            for item in order.items:
                self._session.add(OrderItemModel(
                    order_id=order.order_id,
                    item_id=item.item_id,
                    name=item.name,
                    price=item.price,
                    quantity=item.quantity,
                ))
        await self._session.flush()

    async def delete(self, order_id: str) -> None:
        now = datetime.utcnow()
        await self._session.execute(
            update(OrderItemModel)
            .where(OrderItemModel.order_id == order_id, OrderItemModel.deleted_at.is_(None))
            .values(deleted_at=now)
        )
        await self._session.execute(
            update(OrderModel)
            .where(OrderModel.order_id == order_id, OrderModel.deleted_at.is_(None))
            .values(deleted_at=now)
        )

    async def _load_items(self, order_id: str) -> list[OrderItem]:
        stmt = select(OrderItemModel).where(
            OrderItemModel.order_id == order_id,
            OrderItemModel.deleted_at.is_(None),
        )
        rows = (await self._session.execute(stmt)).scalars().all()
        return [OrderItem(r.item_id, r.name, r.price, r.quantity) for r in rows]

    def _to_domain(self, row: OrderModel, items: list[OrderItem]) -> Order:
        return Order(
            order_id=row.order_id,
            user_id=row.user_id,
            items=items,
            status=row.status,  # type: ignore[arg-type]
        )
